### ISSUE TITLE
L2 implementation difference between keras and caffe

### ISSUE THREAD

**Issue by Queequeg92** [fchollet/keras#4712]
When I set L2 to some typical values(e.g. 0.005) in some state-of-the-art papers(usually use deep networks with lots of parameters), I got a much higher training loss compared with caffe. So I cannot repeat the results of same state-of-the-art papers(implemented in caffe) using the same hyperparameters in keras. Does anyone can explain the L2 implementation difference  or other differences between keras and caffe?

**Comment by the-moliver** [comment]
Keras includes the contribution of the regularizers in the reported training loss (which would make it higher), while it would seem Caffe does not. Caffe implements the regularization by modifying the gradients of the loss function, while Keras implements the regularization by adding the penalty to the loss function and letting automatic differentiation take care of modifying the gradients of the loss function. I believe the effective penalty on the weights implemented in Keras will be twice as strong for the same hyperparameter value compared to Caffe due to taking the derivative of the squared  values, but otherwise should behave similarly.

**Comment by Queequeg92** [comment]
@the-moliver Thanks for your answer. I have checked the source code of Keras. But I think the equivalent L2 in Keras should be half of the same hyperparameter value in Caffe, which producing the same differentiation as in Caffe. I will try the equivalent hyperparameters in keras and test if they works.

**Comment by the-moliver** [comment]
Yes, you would use half the hyperparameter in Keras as opposed to Caffe, because the update to the gradient would be twice as strong for the same hyperparameter, as I said.
